text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+
+### Extends
+
+Using the `pnpm.templates.extends` mechanism, `package.json` fields from templates are copied with small exceptions. The `name`, `dist`, and underscore prefixed fields (e.g. `_npmUser`) are not copied since they're typically specific to the template package itself.
+
+A package referencing the `@example/react-template` above will have the following on-disk and in-memory representations.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "extends": ["@example/react-template@0.1.0"],
+    }
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "version": "0.1.0",
+  "author": "Example Team <team@team.example>",
+  "license": "MIT",
+  "scripts": {
+    "compile": "tsc"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+Note that non-standard `package.json` fields such as `eslintConfig` will also be copied. Although these fields have no meaning to package managers, they will appear in the published manifest.
+
+### Toolkit
+
+The `pnpm.templates.toolkit` field copies the `dependencies` block of the template and spreads it into the `devDependencies` of the referencing package.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "toolkit": ["@example/react-template@0.1.0"],
+    }
+  },
+  "devDependencies": {
+    "jest": "^29.4.3"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "redux": "^4.2.1"
+  },
+  "devDependencies": {
+    "jest": "^29.4.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+This allows package authors to share a reusable set of tools that can be imported or required in the referencing package. Dependencies of dependencies are not normally visible in this manner since pnpm sets up a semi-strict `node_modules` structure by default.
+
+### Catalog
+
+The `pnpm.templates.catalog` feature allows packages to declare a dependency using a version specifier from the referenced template. The `dependencies` block of the template will be available to reference through the `catalog:` version specifier protocol.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "catalog": ["@example/react-template@0.1.0"],
+    }
+  },
+  "dependencies": {
+    "react": "catalog:"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "react": "^18.2.0"
+  }
+}
+```
+
+We expect most monorepos to use this feature to keep dependency specifiers consistent between different in-repo packages.
+
+### Combining Templates
+
+Templates may not refer to other templates. Instead of an inheritance hierarchy, a package may refer to multiple templates with later entries in the list taking precedence.
```

Comment:
ok, but I guarantee that people will ask for inheritance. We should probably suggest some workarounds to them. Like they could use a prepublish script in the template to copy fields from another template.

Reply:
I was thinking few people would ask for inheritance, but I think you have a much better understanding of this side of the ecosystem than I do. Let's definitely consider it. 🙂

Reply:
I don't know if many people will ask for it but when there is a "missing feature", someone always asks for it.

Reply:
If we spell it out "This is why there is no inheritance" in documentation ahead of time, it's an easy point-to for anyone who might ask. Especially if we state something like "We understand people may ask for this feature but it's not happening"

Reply:
Copying your other comment @shellscape:

> Agreed with "all" being OK. Different templates could extend different concerns. If templates can be composed via extends then we have what we need.

This was what I was imagining too. Multiple small templates that have different concerns/purposes, but can be composed together. If that ideal is met, I'm not certain if users will ask for inheritance after documentation can explain the template composition mental model.

But at the moment I'm not opposed to template to template inheritance if @zkochan is in favor of it.